### PR TITLE
Fixes #18736 - Forced unlimited check for filter with no tax

### PR DESCRIPTION
**Pull request by dhlavac**

### CODE REVIEW

File: app/models/role.rb [hunk 1]
```diff
@@ -232,7 +232,7 @@ def self.override_search_operator(operator)
   private
 
   def sync_inheriting_filters
-    self.filters.where(:override => false).each { |f| f.inherit_taxonomies! }
+    self.filters.where(:override => false).each { |f|  f.save }
```
Comment (ares) [inline app/models/role.rb]: Seems like extra space before f.

**Review by ares — CHANGES_REQUESTED** [review]
Would you mind adding the test for the original issue that updates to role, verifying the filter was updated too?

**Comment by dhlavac** [comment]
@ares Thank you  for review - Changes done please look at that

**Comment by dhlavac** [comment]
[test]

File: app/models/role.rb [hunk 2]
```diff
@@ -232,7 +232,11 @@ def self.override_search_operator(operator)
   private
 
   def sync_inheriting_filters
-    self.filters.where(:override => false).each { |f| f.inherit_taxonomies! }
+    filters_check = []
+    self.filters.where(:override => false).each { |f| filters_check.push(f.save) }
+    errors.add :base, N_('Some of associated filters is invalid which prevented the role to be saved') if filters_check.include? false
+    errors.empty?
+    raise ActiveRecord::Rollback, N_("Unable to submit role: Problem with associated filter") if filters_check.include? false
```
Comment (dLobatog) [inline app/models/role.rb]: How about showing the error by finding the filter that failed then printing `#{filter.errors}`?

File: test/models/role_test.rb [hunk 3]
```diff
@@ -315,6 +315,19 @@ def test_allow_to_modify_roles_when_bypass_locking
...
+      it 'should forced unlimited check if role or filter have no taxonomies' do
+        @role.organizations = [ @org1 ]
+        @role.save
+        @filter_with_org.reload
+        assert @role.filters.last.unlimited?
+        refute @role.filters.first.unlimited?
+        @role.organizations = []
+        @role.save
+        @filter_with_org.reload
+        assert @role.filters.last.unlimited?
```
Comment (dLobatog) [inline test/models/role_test.rb]: Not sure if the assertion for `.last` means anything?

File: app/models/role.rb [hunk 4]
```diff
@@ -232,7 +232,11 @@ def self.override_search_operator(operator)
   private
 
   def sync_inheriting_filters
-    self.filters.where(:override => false).each { |f| f.inherit_taxonomies! }
+    filters_check = []
+    self.filters.where(:override => false).each { |f| filters_check.push(f.save) }
+    errors.add :base, N_('Some of associated filters is invalid which prevented the role to be saved') if filters_check.include? false
```
Comment (dLobatog) [inline app/models/role.rb]: `One or more of the associated filters are invalid: #{error}` - it's more similar to other errors in Foreman and also gives a hint to the user about what to do.

**Review by dLobatog — CHANGES_REQUESTED** [review]
Thanks @dhlavac - left a few comments

File: test/models/role_test.rb (hunk 3 above)
Comment (dhlavac) [reply test/models/role_test.rb]: In this case when we have 2 filters in role  I only test if something happend with filter with unlimited check by default - after save .. e.g. unlimited check unchcecked

**Comment by dhlavac** [comment]
@dLobatog Thanks for review changes done and left one comment please look at that

**Comment by dhlavac** [comment]
@dLobatog @ares Changes done.

File: test/models/role_test.rb (hunk 3 above)
Comment (ares) [reply test/models/role_test.rb]: I think you could use `@filter_without_org` and `@filter_with_org` variables that are defined above. That deduplicates the logic.

File: app/models/role.rb [hunk 5]
```diff
@@ -232,7 +232,13 @@ def self.override_search_operator(operator)
   private
 
   def sync_inheriting_filters
-    self.filters.where(:override => false).each { |f| f.inherit_taxonomies! }
+    self.filters.where(:override => false).each do |f|
+      unless f.save
+        errors.add :base, N_('One or more of the associated filters are invalid which prevented the role to be saved')
+        errors.empty?
```
Comment (ares) [inline app/models/role.rb]: What does this line do? It seems like it has no effect?

**Review by ares — CHANGES_REQUESTED** [review]
While the added test is good to have, it does not cover the reason for the fix I think. Could you please add a another test where a filter would be invalid and it would check that role was not saved (save was rollbacked) and contains the validation error? Something like

```ruby
refute role_with_invalid_filters.save
asser rile_with_invalid_filter.errors[:base].includes "...filters are invalid..."
```

File: app/models/role.rb (hunk 5 above)
Comment (dhlavac) [reply app/models/role.rb]: How can I create invalid filter ? -> filter with no permissions doesnt work

**Comment by dhlavac** [comment]
@ares How can I create invalid filter ? -> filter with no permissions doesn't work

**Comment by ares** [comment]
> @ares How can I create invalid filter ? -> filter with no permissions doesn't work

looking at filter model, there are several validate definitions that could be broken, e.g. setting search to empty string, assigning 2 permissions each for different resource type, role set to nil, linked role is locked. Some of it should work I hope.

**Comment by dhlavac** [comment]
@ares I cannot find way to create invalid filter

**Comment by ares** [comment]
@dhlavac 

how about this?
```ruby
# if you need unsaved
FactoryGirl.build(:filter, :permissions => Permission.where(:name => ['view_domains', 'view_subnets']))

# if you need persisted
f = $some_filter
f.permissions = Permission.where(:name => ['view_domains', 'view_subnets'])
f.save(:validate => false)
```

**Review by dLobatog — APPROVED** [review]
Thanks @dhlavac

**Pull request merged**